pandas/core/arrays/datetimelike.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from pandas._libs import iNaT
+from pandas._libs.tslibs.timedeltas import delta_to_nanoseconds
+
+from pandas.tseries import frequencies
+
+import pandas.core.common as com
+from pandas.core.algorithms import checked_add_with_arr
+
+
+class DatetimeLikeArrayMixin(object):
+    """
+    Shared Base/Mixin class for DatetimeArray, TimedeltaArray, PeriodArray
+
+    Assumes that __new__/__init__ defines:
+        _data
+        freq
+    """
+
+    @property
+    def _box_func(self):
+        """
+        box function to get object from internal representation
+        """
+        raise com.AbstractMethodError(self)
+
+    def __iter__(self):
+        return (self._box_func(v) for v in self.asi8)
+
+    @property
+    def values(self):
+        """ return the underlying data as an ndarray """
+        return self._data.view(np.ndarray)
```

Comment:
Do we want this for our arrays? 
I am not sure this should be called `.values`. 
And we already have `_ndarray_values` ?

Reply:
This is a good question, and I think relates to the inheritance vs composition question.  As with several other comments, the narrow-context answer is that this PR is designed to involve essentially zero _changes_ to behavior, is just moving methods/properties from their existing locations in the index classes.

As for questions about whether a method/property is needed in the array classes, these are all going to be needed by the arithmetic/comparison methods (ported in a later pass)